xapian-core/cluster/kmeans.cc

```diff
@@ -0,0 +1,142 @@
+/** @file kmeans.cc
+ *  @brief KMeans clustering API
+ */
+/* Copyright (C) 2016 Richhiey Thomas
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian/cluster.h"
+#include <debuglog.h>
+#include <xapian/error.h>
+
+#include <cstdlib>
+#include <climits>
+#include <set>
+#include <unordered_map>
+#include <vector>
+
+#define EPSILON 0.0000000001
+
+using namespace Xapian;
+using namespace std;
+
+string
+KMeans::get_description() const {
+    LOGCALL(API, string, "KMeans::get_description()", NO_ARGS);
+    return "KMeans clusterer";
+}
+
+void
+KMeans::initialize_centroids(ClusterSet &cset) {
+    LOGCALL_VOID(API, "KMeans::initialize_random()", cset);
+    set<docid> centroid_docs;
+    int size = docs.size();
+    srand(time(NULL));
+    for (unsigned int i = 0; i < k;) {
+	int x = rand() % size;
+	if (centroid_docs.find(x) == centroid_docs.end()) {
+	    Point a = docs[x];
+	    Centroid centroid;
+	    centroid.set_to_point(a);
+	    centroids.push_back(centroid);
+	    centroid_docs.insert(x);
+	    Cluster cluster1(centroid);
+	    cset.add_cluster(cluster1);
+	    ++i;
+	}
+    }
+}
+
+void
+KMeans::initialize_points(const MSet &source, TermListGroup &tlg) {
+    LOGCALL_VOID(API, "KMeans::initialize_points()", source | tlg);
+    for (MSetIterator it = source.begin(); it != source.end(); ++it) {
+	Point p;
+	p.initialize(tlg, it.get_document());
+	docs.push_back(p);
+    }
+}
+
+bool
+KMeans::converge(vector<Centroid> &previous, vector<Centroid> &current) const {
+    LOGCALL(API, bool, "KMeans::converge()", previous | current);
+    CosineDistance cosine;
+    for (unsigned int i = 0; i < k; ++i) {
+	Centroid &prev = previous[i];
+	Centroid &curr = current[i];
+	double dist = cosine.similarity(prev, curr);
+	if (dist > EPSILON)
+	    return false;
+    }
+    return true;
+}
+
+ClusterSet
+KMeans::cluster(MSet &mset) {
+    LOGCALL(API, ClusterSet, "KMeans::cluster()", mset);
+    unsigned int size = mset.size();
+    mset.fetch();
+    if (size <= 0)
```

Comment:
> I wonder if it's more helpful to treat clustering an empty MSet as a valid operation

That's the way I have kept it currently, since clustering an empty MSet will only return an empty ClusterSet. I guess the better way to solve this would be:
1) In an empty MSet case, let it not be handled separately and return an empty CSet as it would
2) In a size <= k case, we could make k = size and return the CSet in which every doc has its own cluster.
I'm still not sure whether this is useful, because in practice, the value of **_size_** will be greater than the value of **_k_** by a fair margin. Therefore, this case could be seen as a mistake on the users part and an error would be more helpful than a wrongly clustered ClusterSet